photosynthesis_metrics/inception_score.py

```diff
@@ -0,0 +1,94 @@
+
+    Returns:
+        score, variance:
+
+    Reference:
+        https://arxiv.org/pdf/1801.01973.pdf
+    """
+    N = features.size(0)
+    # Convert logits to probabilities
+    probas = F.softmax(features)
+    # In paper score computed for 10 splits of dataset and then averaged.
+    partial_scores = []
+    for i in range(num_splits):
+        subset = probas[i * (N // num_splits): (i + 1) * (N // num_splits), :]
+        # Compute KL divergence
+        p_y = torch.mean(subset, dim=0)
+        scores = []
+        for k in range(subset.shape[0]):
+            p_yx = subset[k, :]
+            scores.append(F.kl_div(p_y.log(), p_yx, reduction='sum'))
+        # Compute exponential of the mean of the KL-divergence for each split
+        partial_scores.append(torch.tensor(scores).mean().exp())
+
+    partial_scores = torch.tensor(partial_scores)
+    return torch.mean(partial_scores).to(features), torch.std(partial_scores).to(features)
+
+
+class IS(BaseFeatureMetric):
+    r"""Creates a criterion that measures difference of Inception Score between two datasets.
+    IS is computed separatly for predicted and target features and expects raw InceptionV3 model logits as inputs.
+
+    Args:
+        predicted_features (torch.Tensor): Low-dimension representation of predicted image set.
+            Required to have shape (N_pred, encoder_dim)
+        target_features (torch.Tensor): Low-dimension representation of target image set.
+            Required to have shape (N_targ, encoder_dim)
+
+    Returns:
+        distance(predicted_score, target_score): L1 or L2 distance between scores.
+
+    Reference:
+        https://arxiv.org/pdf/1801.01973.pdf
+    """
+    def __init__(self, num_splits: int = 10, distance: str = 'l1') -> None:
+        r"""
+        Args:
+            num_splits: Number of parts to devide features.
+                IS is computed for them separatly and results are then averaged.
+            distance: How to measure distance between scores. One of {`l1`, `l2`}. Default: `l1`.
+        """
+        super(IS, self).__init__()
+        self.num_splits = num_splits
+        self.distance = distance
+
+    def compute_metric(
+            self, predicted_features: torch.Tensor, target_features: torch.Tensor) -> torch.Tensor:
+        r"""Compute IS
+
+        Returns:
+            predicted_score: Scalar value of IS for predicted images features.
+            target_score: Scalar value of IS for target images features if `ret_target` is True.
+        """
+        predicted_is, _ = inception_score(predicted_features, num_splits=self.num_splits)
+        target_is, _ = inception_score(target_features, num_splits=self.num_splits)
+        if self.distance == 'l1':
+            return torch.dist(predicted_is, target_is, 1)
+        elif self.distance == 'l2':
+            return torch.dist(predicted_is, target_is, 2)
```

Comment:
What happens if `self.distance` is neither 'l1' nor 'l2'? I think there should be `else` with `ValueError` and error message that explains that invalid distance type is provided